src/main/java/jenkins/plugins/openstack/compute/slaveopts/BootSource.java

```diff
@@ -346,6 +347,21 @@ public String toString() {
             return "Volume from Image " + name + " (" + volumeSize + "GB)";
         }
 
+        @Override
+        public boolean equals(Object o) {
+            if (this == o)
+                return true;
+            if (o == null || getClass() != o.getClass() || !super.equals(o))
```

Comment:
The parent class's `equals` method includes the code `if (o == null || getClass() != o.getClass()) return false;` so this line of code does not _need_ to duplicate that functionality - it could be replaced simply by `if (!super.equals(o))` and be functionally equivalent.

FYI if you ask the Eclipse IDE (and probably other IDEs) to generate an `equals` method for a class which has `super.equals` then it'll generate code of the form
```
@Override
public boolean equals(Object obj) {
    if (this == obj) {
        return true;
    }
    if (!super.equals(obj)) {
        return false;
    }
    ...
```

I'm open to discussion on whether or not you (and/or anyone else) agree that making such a change leads to good coding style or not (different folks have different opinions :grin:) but, personally, my feeling is that, all else being equal, less/smaller code is preferable to more/larger code.

Reply:
Almost agree, right now you are 100% correct. But what if the impl. of super changes? Personally I'm not fan to be so dependant on the super, but no strong preferrence here.
Anyway both way are correct, so let's make it...

Reply:
If super.equals changes then that'll be caught by the unit test, right? :grin:

Seriously tho', unit tests are the answer; if you're worried that someone might bork the superclass and, by implication, bork the child class, then adding some unit tests that verify that hashcode and equals operate correctly for both these classes is the best way to ensure that this doesn't happen accidentally.
I'll leave it up to you to decide if you're worried enough to write some unit tests :wink: 

FYI, historically, when I originally wrote the VolumeSnapshot support code (in #136), `BootSource` was an enum (so didn't need these), and then (after lengthy discussions) it was refactored to match the same pattern as `LauncherFactory`; `LauncherFactory` had `equals` and `hashcode` methods, so I had my IDE generate some for my `BootSource` code, resulting in `BootSource.Image` and `BootSource.VolumeSnapshot` having these methods but, due to coding fatigue at that point (by then, it was over 6 months since I'd started on this work), I failed to write unit tests for these methods.
`BootSource.VolumeFromImage` was added later (#194 and #195), but without `equals` and `hashcode` methods and so that didn't add any unit-tests either.

Reply:
Relying on supper implementation is fine. Subclass is not here to correct possible superclass' mistakes. If the superclass implementation changes, it is mostly because it private state representation has changed which should be hidden from the subclass and relying on result of `equals` permits such decoupling.